### PR TITLE
Allow out-of-bound fence requests

### LINKED ISSUE #116: Allow fences to be out of bounds
The current behaviour is to throw if any of the input coordinates are out of bounds. However that might be too restrictive for some. And alternative approach is to use a fillValue for traces that are out of bounds, similar to how we plan to handle horizons #114

### PR DESCRIPTION
**Pull request by AdilKhurshid**
The current behaviour is to throw an error if any coordinate provided in the request falls outside the boundaries of the seismic volume. However, with this pull request, the user will now be able to submit requests containing such coordinates and the response will return fillvalues for all values lying outside the seismic cube.

closes #116

### CODE REVIEW

File: internal/core/cppapi_data.cpp [hunk 1]
```diff
@@ -189,14 +193,8 @@ void fence(
...
         auto validate_boundary = [&] (const int voxel, Axis const& axis) {
-            if(!axis.inrange(coordinate[voxel])){
-                const std::string coordinate_str =
-                    "(" +std::to_string(x) + "," + std::to_string(y) + ")";
-                throw std::runtime_error(
-                    "Coordinate " + coordinate_str + " is out of boundaries "+
-                    "in dimension "+ std::to_string(voxel)+ "."
-                );
-            }
+            if(!axis.inrange(coordinate[voxel]))
```
Comment (achaikou) [inline internal/core/cppapi_data.cpp:196]: I think using { } is always safer for "ifs". 
From experience it is too easy to not notice when adding second line, for example for debugging, and then wonder why my code doesn't work the way I want it to :smile:

File: internal/core/cppapi_data.cpp [hunk 2]
```diff
@@ -218,6 +216,7 @@ void fence(
         interpolation_method
     );
 
+    write_fillvalue(data.get(), noval_indicies, nsamples, fillValue);
```
Comment (achaikou) [inline internal/core/cppapi_data.cpp:219]: Wonder if we should also rename the third agrument here, in additional commit.
It is no longer "vertical_size", it gives no meaning here.

File: api/request.go [hunk 3]
```diff
@@ -98,6 +98,10 @@ type FenceRequest struct {
 	// Note: For nearest interpolation result will snap to the nearest point
 	// as per "half up" rounding. This is different from openvds logic.
 	Interpolation string `json:"interpolation" example:"linear"`
+
+	// FillValue will be used for any sample point in the fence that falls
+	// outside the bounds of the seismic volume.
+	FillValue *float32 `json:"fillValue" binding:"required"`
```
[the remark this answers is not in the document]
Comment (ErlendHaa) [reply api/request.go:104]: Sveinung had a good suggestion for once. Make it optional, if no fill value is provided then we preserve the old behaviour. I.e. we throw out of bounds. But if the users comes with a fill value, then we populate with that in stead of throwing. 

This has the added benefit that it doesn't break the API

File: internal/core/cppapi_data.cpp [hunk 4]
```diff
@@ -178,9 +181,10 @@ void fence(
             }
         }
     };
-
-    Axis inline_axis = metadata.iline();
+    Axis inline_axis    = metadata.iline();
     Axis crossline_axis = metadata.xline();
+    Axis samples         = metadata.sample();
```
[the remark this answers is not in the document]
Comment (ErlendHaa) [reply internal/core/cppapi_data.cpp:186]: And if you decide to vertically align with the above lines, then it should align 😄

File: internal/core/core_test.go [hunk 5]
```diff
@@ -18,9 +18,10 @@ func make_connection(name string) Connection {
 	return NewFileConnection(path)
 }
 
-var well_known = make_connection("well_known")
-var samples10 = make_connection("10_samples")
-var prestack = make_connection("prestack")
+var   well_known = make_connection("well_known")
+var   samples10  = make_connection("10_samples")
+var   prestack   = make_connection("prestack")
+const fillValue  = float32(-999.25)
```
Comment (ErlendHaa) [inline internal/core/core_test.go:24]: I'm not a fan of formatting unrelated code like this. I highly encourage you do to these kinds of small changes when you see them, but always as separate commits

File: api/request.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (AdilKhurshid) [reply api/request.go:104]: Sure, I will make this an optional field and I will mention this new behavior of the fence endpoint in the documentation 👍

File: internal/core/capi.cpp [hunk 6]
```diff
@@ -168,6 +168,7 @@ int fence(
     DataHandle* handle,
     enum coordinate_system coordinate_system,
     const float* coordinates,
+    float fillValue,
```
[the remark this answers is not in the document]
Comment (AdilKhurshid) [reply internal/core/capi.cpp:171]: Sure

File: tests/e2e/test_requests.py [hunk 7]
```diff
@@ -46,12 +46,19 @@ def make_slice_request(vds=VDSURL, direction="inline", lineno=3, sas="sas"):
...
+        fillValue=-999.25
+    ):
     return {
         "vds": vds,
         "coordinateSystem": coordinate_system,
         "coordinates": coordinates,
-        "sas": sas
+        "sas": sas,
+        "fillValue": fillValue,
     }
```
[the remark this answers is not in the document]
Comment (AdilKhurshid) [reply tests/e2e/test_requests.py:63]: I'm making fill value an optional field then in that case it will not affect our performance test suits

File: internal/core/core_test.go [hunk 8]
```diff
@@ -459,51 +461,61 @@ func TestFence(t *testing.T) {
...
+			crd_system : CoordinateSystemCdp,
+			coordinates: [][]float32{{700, 1200}},
+			fence      : []float32{-999.25, -999.25, -999.25, -999.25},
 		},
 		{
-			name:              "coordinate 2 is just-out-of-lower-bound in direction 1",
-			coordinate_system: CoordinateSystemAnnotation,
-			coordinates:       [][]float32{{0, 11}, {5.9999, 10}, {0.0001, 9.4999}},
-			err:               "is out of boundaries in dimension 1.",
+			name       : "coordinate 2 is just-out-of-lower-bound in direction 1",
```
[the remark this answers is not in the document]
Comment (AdilKhurshid) [reply internal/core/core_test.go:489]: I will keep it the same for this commit and add an additional commit message to fixing this.

File: internal/core/cppapi_data.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (AdilKhurshid) [reply internal/core/cppapi_data.cpp:186]: sure

File: internal/core/cppapi_data.cpp (hunk 1 above)
Comment (AdilKhurshid) [reply internal/core/cppapi_data.cpp:196]: Sure, 😂

File: internal/core/cppapi_data.cpp (hunk 2 above)
Comment (AdilKhurshid) [reply internal/core/cppapi_data.cpp:219]: sure

File: internal/core/core_test.go (hunk 5 above)
Comment (AdilKhurshid) [reply internal/core/core_test.go:24]: Sure

File: internal/core/cppapi_data.cpp [hunk 9]
```diff
@@ -189,13 +193,17 @@ void fence(
...
+            if (!axis.inrange(coordinate[voxel])) {
+                if (fillValue != nullptr) {
+                    noval_indicies.push_back(i * nsamples);
+                } else {
+                    const std::string coordinate_str =
+                        "(" +std::to_string(x) + "," + std::to_string(y) + ")";
+                    throw std::runtime_error(
+                        "Coordinate " + coordinate_str + " is out of boundaries "+
+                        "in dimension "+ std::to_string(voxel)+ "."
+                    );
```
Comment (ErlendHaa) [inline internal/core/cppapi_data.cpp:205]: You can reduce the nesting here by swapping the if comparison:

```
if (fillvalue == nullptr) {
    throw ...
}

noval_indicies.push_back(i * nsamples)
```

File: api/request.go [hunk 10]
```diff
@@ -99,9 +97,11 @@ type FenceRequest struct {
 	// as per "half up" rounding. This is different from openvds logic.
 	Interpolation string `json:"interpolation" example:"linear"`
 
-	// FillValue will be used for any sample point in the fence that falls
-	// outside the bounds of the seismic volume.
-	FillValue *float32 `json:"fillValue" binding:"required"`
+	// Providing a FillValue is optional and will be used for the sample point
+	// that lies outside the seismic cube.
```
[the remark this answers is not in the document]
Comment (AdilKhurshid) [reply api/request.go:101]: agree

File: internal/core/cppapi_data.cpp [hunk 11]
```diff
@@ -193,8 +193,18 @@ void fence(
...
         auto validate_boundary = [&] (const int voxel, Axis const& axis) {
-            if(!axis.inrange(coordinate[voxel]))
-                noval_indicies.push_back(i * nsamples);
+            if (!axis.inrange(coordinate[voxel])) {
+                if (fillValue != nullptr) {
+                    noval_indicies.push_back(i * nsamples);
+                } else {
+                    const std::string coordinate_str =
+                        "(" +std::to_string(x) + "," + std::to_string(y) + ")";
+                    throw std::runtime_error(
```
[the remark this answers is not in the document]
Comment (AdilKhurshid) [reply internal/core/cppapi_data.cpp:202]: I kept it the same as it was like this earlier. Yes, I agree we should return bad requests in this case, but the question is should we add that in the Golang layer or here?

File: internal/core/core_test.go [hunk 12]
```diff
@@ -515,12 +516,73 @@ func TestFenceBorders(t *testing.T) {
...
 		handle, _ := NewVDSHandle(well_known)
 		defer handle.Close()
-		_, err := handle.GetFence(testcase.coordinate_system, testcase.coordinates, interpolationMethod)
+		_, err := handle.GetFence(testcase.coordinate_system, testcase.coordinates, interpolationMethod, nil)
 
 		require.ErrorContainsf(t, err, testcase.err, "[case: %v]", testcase.name)
 	}
 }
 
+func TestFenceBordersWithFillValue(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (AdilKhurshid) [reply internal/core/core_test.go:525]: I think combining both tests will increase the complexity of the test case. I will change the order of the commit before merging so that it makes more sense.

File: internal/core/cppapi_data.cpp (hunk 11 above)
[the remark this answers is not in the document]
Comment (ErlendHaa) [reply internal/core/cppapi_data.cpp:202]: If you throw detail::bad_request here, the correct error type will be propagated correctly all the way back to the user. I.e. detail::bad_request ultimately ends up as a HttpBadRequest

[the remark this answers is not in the document]
Comment (achaikou) [reply internal/core/cppapi_data.cpp]: In theory it would have been better to make this change in its own commit, but I can accept it being together as it is relevant to the changed code...

**Review by achaikou — COMMENTED** [review]
In my opinion can prepare for merge :cat:

File: api/request.go [hunk 13]
```diff
@@ -97,7 +97,7 @@ type FenceRequest struct {
 	// as per "half up" rounding. This is different from openvds logic.
 	Interpolation string `json:"interpolation" example:"linear"`
 
-	// Providing a FillValue is optional and will be used for the sample point
+	// Providing a FillValue is optional and will be used for the sample points
 	// that lies outside the seismic cube.
```
[the remark this answers is not in the document]
Comment (AdilKhurshid) [reply api/request.go:101]: It's completely fine to torture me, It's good for my learning 😊

[the remark this answers is not in the document]
Comment (AdilKhurshid) [reply api/request.go]: it's done 
https://github.com/equinor/vds-slice/pull/186/files#diff-a6c4e8b234cb1efabeebf2d063fc999ab11953a19c35d2f92063f542cff7b328R101

**Review by achaikou — APPROVED** [review]
Looks good at least to me. See if you want to change last comment and if Erlend has something more to add.
